lib/SILOptimizer/Transforms/PerformanceInliner.cpp

```diff
@@ -351,6 +349,11 @@ bool SILPerformanceInliner::isProfitableToInline(FullApplySite AI,
   int BaseBenefit = RemovedCallBenefit;
 
   SubstitutionMap CalleeSubstMap;
+  if (IsGeneric) {
+    CalleeSubstMap = Callee->getGenericEnvironment()
+      ->getSubstitutionMap(AI.getSubstitutions());
```

Comment:
Once the new representation is done, going from a list to a map will be O(1), and going from a map to a list will just be a memcpy :-)